controller/src/main/java/io/pravega/controller/server/bucket/BucketManager.java

```diff
@@ -0,0 +1,190 @@
+/**
+ * Copyright (c) 2017 Dell Inc., or its subsidiaries. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.pravega.controller.server.bucket;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.util.concurrent.AbstractService;
+import io.pravega.common.concurrent.Futures;
+import io.pravega.controller.store.stream.BucketStore;
+import lombok.AccessLevel;
+import lombok.Getter;
+import lombok.extern.slf4j.Slf4j;
+
+import javax.annotation.concurrent.GuardedBy;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Executor;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+/**
+ * This class is central to bucket management. This is instantiated once per service type. It is responsible for acquiring
+ * ownership of buckets for the given service and then starting one bucket service instance for each bucket it owns. 
+ * It manages the lifecycle of bucket service.
+ * The implementation of this class implements store specific monitoring for bucket ownership changes.
+ * This service does not have any internally managed threads. Upon receiving any new notification, it submits a "tryTakeOwnership" 
+ * work in the executor's queue. 
+ */
+@Slf4j
+public abstract class BucketManager extends AbstractService {
```

Comment:
Bucket manager manages bucket services lifecycle. 
It is abstract because it comes in two flavours -- zookeeper bucket manager and in-memory bucket manager. 
It is also instantiated on a per service basis because we could choose different bucketing logic for different services. We currently have `RetentionService` and we will add `WatermarkingService` in future. 
The work that is supplied and performed at a certain frequency for bucketService will vary for these. 
Streams added for retention may be completely different from streams added for watermarking. So each service type would have its own sets of buckets where streams are added according to what capability is enabled. This allows us to keep number of buckets different for different service types too.